Unity-Technologies/Graphics
# Dev aov light decomposition

Purpose of this PR
--------------------
This PR adds new debugging lighting modes to be used for light decomposition AOVs.
In the offline rendering light decomposition AOV are often used by the compositors to re-light scenes. By adding these new debug modes we will allow to export these render passes.
lighting_debug_modes
![light_decomposition](https://user-images.githubusercontent.com/35697738/78951845-8650ff80-7aa1-11ea-834c-e8867e52237f.gif)

The new DebugLightingMode added are DirectDiffuse, DirectSpecular, IndirectDiffuse, Reflection,
Refraction, Transmittance and Emissive.

The light loop is modified to output the decomposed lighting, the decomposed lighting is saved into 19 floats. This data is then used in HDRP lighting debug system.
The performance impact of this on HDRP non-debug (run-time) shader should not be significant because it is only used when in debug mode,
During shader compilation the shader compiler should remove the unnecessary instructions from HDRP non-debug (run-time) shader.

This PR also contain a change to the method debugDisplaySettings in HDRenderPipeline, I modified this method to make it refer to the m_CurrentDebugDisplaySettings
because the m_DebugDisplaySettings can be overriden by neutral debug display settings values and debug setting values to create AOVs
see aovRequest.SetupDebugData

Here are the detailed changes:

HDRP/Runtime/Debug/DebugDisplay.cs
-->Add debugLighting == DebugLightingMode.DirectDiffuse ||
debugLighting == DebugLightingMode.DirectSpecular ||
debugLighting == DebugLightingMode.IndirectDiffuse ||
debugLighting == DebugLightingMode.DiffuseLighting ||
debugLighting == DebugLightingMode.Reflection||
debugLighting == DebugLightingMode.Refraction ||
To DebugNeedsExposure

Runtime/Debug/LightingDebug.cs
Added new DebugLightMode DirectDiffuse, DirectSpecular, IndirectDiffuse, Reflection, Refraction, Transmittance, Emissive

Runtime/Debug/LightingDebug.cs.hlsl

Adding new define for the new debug lighting modes

Runtime/Lighting/Deferred.shader
Runtime/Lighting/LightLoop/Deferred.compute
Runtime/Lighting/LightLoop/DeferredTile.shader
Added new call DecomposedLighting decomposedLighting
and PostLightLoopDebugDisplayCall to fill in the decomposed light data after the light loop

Runtime/Lighting/LightLoop/LightLoop.hlsl
Added out argument DecomposedLighting to the LightLoop

Runtime/Material/MaterialEvaluation.hlsl
new struct defined

struct DecomposedLighting
{
float3 directDiffuse;
float3 directSpecular;
float3 indirectDiffuse;
float3 reflection;
float3 refraction;
float transmittance;
float3 emissive;
};

Runtime/Material/AxF/AxF.hlsl
Runtime/Material/Eye/Eye.hlsl
Runtime/Material/Fabric/Fabric.hlsl
Runtime/Material/Hair/Hair.hlsl
Runtime/Material/Lit/Lit.hlsl
Runtime/Material/Lit/SimpleLit.hlsl
Runtime/Materila/Lit/StackLit.hlsl
Added out parameter DecomposedLighting to the PostEvaluateBSDF function
Fill in the decomposed lighting data

Runtime/RenderPipeline/HDRenderPipeline.cs
Disable post processing and enable exposure for rendering aovs that requires them
in ExecuteRenderRequest

Runtime/RenderPipeline/Raytracing/Shaders/RayTracingLightCluster.hlsl
Added out parameter DecomposedLighting to the LightLoop function
and to the PostEvaluateBSDF

Runtime/RenderPipeline/Raytracing/Shaders/Deferred/RaytracingDeferred.compute
Runtime/RenderPipeline/ShaderPass/ShaderPassForward.hlsl
Runtime/RenderPipeline/ShaderPass/ShaderPassRaytracingForward.hlsl
Runtime/RenderPipeline/ShaderPass/ShaderPassRaytracingIndirect.hlsl
Runtime/RenderPipeline/VFXGraph/Shaders/VFXLitPixelOutput.hlsl
Added new out param DecomposedLighting decomposedLighting to LightLoop
and PostLightLoopDebugDisplayCall to fill in the decomposed light data after the light loop

Runtime/Sky/GradientSky/GradientSky.shader
Runtime/Sky/HDRISky/HDRISky.shader
Runtime/Sky/PhysicallyBasedSky/PhysicallyBasedSky.shader
Added call to ModifySkyColorDebug
under the DEBUG_DISPLAY ifdef to turn off sky for debug
display of some AOVs

Runtime/Sky/SkyDebugUtils.hlsl
Add function ModifySkyColorDebug to overide the sky color when DebugLightingMode are enabled

Runtime/RenderPipeline/RenderPass/AOV/AOVRequest.cs
Added new aovs
DiffuseOnly,
SpecularOnly,
DirectDiffuseOnly,
DirectSpecularOnly,
IndirectDiffuseOnly,
ReflectionOnly,
RefractionOnly,
Transmittance,
Enum and call to FillDebugData

Testing status
---------------
Manual Tests: What did you do?

Yes: Opened test project + Run graphic tests locally
Yes: Built a player
[N/A] Checked new UI names with UX convention
[N/A] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
 C# and shader warnings (supress shader cache to see them)
[N/A] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
Other:
Automated Tests: What did you setup? (Add a screenshot or the reference image of the test please)
Will add test on top of the AOV bufer graphics test see PR
2b73cba

Yamato: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline
